### ISSUE TITLE
Multi select component for ui-core

### ISSUE THREAD

**Issue by ismay** [dhis2/notes#58]
Our current `<SelectField>` component does not yet support the `multiple` attribute. We had a talk about how to build it on slack (see [here](https://dhis2.slack.com/archives/CBM8LNEQM/p1564487918000500))

## Background info

* The native multi-select is not that accessible if you don't know how it works, as it needs a keypress to make multiple selections: https://jsfiddle.net/w8qznyc1
* A library that handles multi selects differently than native: https://harvesthq.github.io/chosen (not usable by us though, but a nice example)
* The material-ui multi select: https://material-ui.com/components/selects/#multiple-select
* And over there 👉, on the side, the Github multiselect (see assignees, labels)

In the talk, Joe asked:

> I think that the decision needs to be made: are we staying strict native and adapting DHIS2 UI’s to that, or are we building something custom to DHIS2 needs? Once that decision is made I can work on a proper design spec in collab with you guys

## Proposal

Personally I think that having the SelectField stay close to a native SelectField would be nice. It would keep the code and the component simple. My proposal would be this:

1. Allow a `multiple` attribute on our SelectField. That would then functionally make it work like the [material-ui multi-select](https://material-ui.com/components/selects/#multiple-select), which is closest to our current select-field, and works well I think (though an `X` on a chip to remove it might be nice).
2. For more complicated situations, like when you want filtering, searching, paging, I'd propose a different component, say a `Combobox`. Which is more dynamic, and has a searchfield, etc.

Obviously devs would have to be sensible and not use a SelectField with `multiple` for rendering 10000 items. That's where they'd need to use the ComboBox.

Let me know what you think if you want to partake in the amazing journey of the 📯Dhis2 Multiselect! 📯

**Comment by varl** [comment]
> Allow a multiple attribute on our SelectField. That would then functionally make it work like the material-ui multi-select, which is closest to our current select-field, and works well I think (though an X on a chip to remove it might be nice).

Does this mean that we visually use the same `SelectField` as we do now, but every time an element is selected we add a Chip for it inside the SelectField box?

**Comment by Mohammer5** [comment]
## Standard `select`s

I think it'd be nice to move away from select elements. The UX can be terrible (multi select on both desktop and mobile is not really intuitive).

### Advantages of a native select
The native select has some advantages that I don't want to lose with a custom version
* Works with keyboard inputs when focused
* Works on mobile browsers

### Disadvantages
* Not intuitive when multiple are select-able.
* We can't style the options
* Doesn't work well with a lot of options
* We can't filter the options while keeping them visible

## Another approach
The simplest version I think could be something like this:
![image](https://user-images.githubusercontent.com/10101831/62189831-4bbe9100-b370-11e9-9454-13afe2adf98b.png)

### Advantages
* It would work both single- and multi-select.
* We can add any kind of UI around it
  * chips (for lists with 10k+ items)
  * Search field
  * etc.
* It's fairly easy to control the (max-)height and scrolling (`overflow-y: auto`)
* Display checkboxes (or an icon) to indicate multi-select
  * Or place a small (12px?) sentence above ("You can select multiple items")
* It works both on desktop and mobile browsers
  * We can choose whether we want to wrap white-spaces with a horizontal scrollbar or break options into multiple lines when too long
* Can be used for the group editor component that we'll have to create in `ui-widgets`

### Disadvantages
* it will use more space vertically.

<hr />

Also what @joe wanted to have shouldn't be a problem with this:
![image](https://user-images.githubusercontent.com/10101831/62190356-6f360b80-b371-11e9-8427-c9b32f2bc79a.png)

**Comment by ismay** [comment]
@cooper-joe I need this component in the scheduler-app and wouldn't mind implementing this. How do you feel about @Mohammer5's comment above? Is that something that would work for you? Maybe we need the eyes of the others on this as well before we move ahead?

**Comment by Mohammer5** [comment]
@ismay Please just create a jira issue and assign it to joe, set the one you're working on to 'blocked by' to express the urgency. 

Like @cooper-joe has layed out in [a comment](https://github.com/dhis2/notes/issues/60#issuecomment-525624190) in #60, that's the procedure we should follow in the future:

**Comment by ismay** [comment]
Sure, I’ll create the issue for tracking the work, but I assume @cooper-joe still needs an answer to the question he posed above before he can draft anything. And I assume it’d useful to have any comments before starting the design process.

**Comment by cooper-joe** [comment]
The 'Another approach' proposal laid out in @Mohammer5 's [comment](https://github.com/dhis2/notes/issues/58#issuecomment-516726554) above looks good to me, it covers all requirements. I will work on the design specs this week so that @ismay can implement this while building the scheduler-app.

**Comment by varl** [comment]
Then I think we should also use it for single select (`Select` component) instead of native, and replace the `option` HTML with an `Option` component.

**Comment by joejohncooper** [comment]
@varl Agreed. I'm working on the `select` as we speak, and to me it makes sense that the default component is a single select that can become much more feature-rich if you pass in `multi` for multiple selections, `filter` to add in a search field, and so on. But by default it would be a single select.

**Comment by cooper-joe** [comment]
@varl Agreed. I'm working on the select as we speak, and to me it makes sense that the default component is a single select that can become much more feature-rich if you pass in multi for multiple selections, filter to add in a search field, and so on. But by default it would be a single select.

**Comment by ismay** [comment]
One thing that I'd prefer is if the multi select doesn't include extra features like filtering, etc. Basically, to keep it in ui-core and close to a native multi-select, but with better ux. I'd personally prefer to reserve the additional features for a separate, more complicated component that goes in ui-widgets. Not sure if the others feel that way or not?

**Comment by cooper-joe** [comment]
I propose we close this issue and continue discussion based on the initial designs [here](https://github.com/dhis2/ui-core/issues/406).
